# IMDB Scraping Data

**Start date:** 9/3/2023 

**Scope:** This analysis has the main goal of praticing cleaning data and peforming an exploratory analysis on this dataset.

In [1]:
#Importing the data
# !kaggle datasets download -d bharatnatrayn/movies-dataset-for-feature-extracion-prediction

In [2]:
# import zipfile

# with zipfile.ZipFile('movies-dataset-for-feature-extracion-prediction.zip', 'r') as zip_ref:
#     zip_ref.extractall()

In [3]:
import pandas as pd
import regex as re
import numpy as np

In [4]:
movies_raw_dataset = pd.read_csv('movies.csv')

In [5]:
movies_raw_dataset.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN

On a first glance, there are some issues that will need some work:
- The database does not contain only movies but movies and series
- All text fields contain special characters like '\n'
- RunTime column contains the full length of movies but only the episode length for series
- For beautifying purposes, column names also need to be standardized

Let's start with the columns names and then we have a look at the dataset summary and types

In [6]:
movies_raw_dataset.columns = pd.Series(movies_raw_dataset.columns).apply(lambda x: x[0].upper() + x[1:].lower())

In [7]:
movies_raw_dataset.describe()

Rating      Runtime
count  8179.000000  7041.000000
mean      6.921176    68.688539
std       1.220232    47.258056
min       1.100000     1.000000
25%       6.200000    36.000000
50%       7.100000    60.000000
75%       7.800000    95.000000
max       9.900000   853.000000

Rating column seems fine, as the ratings on IMDB go from 1 to 10. However, 853 for run time seems a little excessive (more than 14 hours). Given that th percentile 75% is 95 minutes, we are likely looking at an outlier here. Let's check it out.

In [8]:
movies_raw_dataset.nlargest(5, columns='Runtime')

Movies         Year  \
1902     El tiempo entre costuras  (2013–2014)   
1081                     Soupçons  (2004–2018)   
2498          The Innocence Files       (2020)   
201    The Haunting of Hill House       (2018)   
820   Cosmos: A Spacetime Odyssey       (2014)   

                                        Genre  Rating  \
1902  \nAdventure, Drama, History                 8.3   
1081  \nDocumentary, Crime, Drama                 7.9   
2498         \nDocumentary, Crime                 8.0   
201      \nDrama, Horror, Mystery                 8.6   
820                 \nDocumentary                 9.3   

                                               One-line  \
1902  \nSira Quiroga is a young Spanish dressmaker e...   
1081  \nThe high-profile murder trial of American no...   
2498  \nCases of wrongful conviction that the Innoce...   
201   \nFlashing between past and present, a fractur...   
820   \nAn exploration of our discovery of the laws ...   

                                                  Stars    Votes  Runtime  \
1902  \n            \n    Stars:\nAdriana Ugarte, \n...    3,876    853.0   
1081  \n            \n    Stars:\nMichael Peterson, ...   20,200    629.0   
2498  \n            \n    Stars:\nPeter Neufeld, \nB...    2,335    573.0   
201   \n            \n    Stars:\nMichiel Huisman, \...  195,117    572.0   
820   \n            \n    Stars:\nNeil deGrasse Tyso...  114,386    557.0   

     Gross  
1902   NaN  
1081   NaN  
2498   NaN  
201    NaN  
820    NaN

So apparently we have mixed values on the RunTime column, with some series having the duration per episode and other having the total duration of the series. We will have to find a way to differentiate these. 

In [9]:
movies_raw_dataset.dtypes

Movies       object
Year         object
Genre        object
Rating      float64
One-line     object
Stars        object
Votes        object
Runtime     float64
Gross        object
dtype: object

Both votes and Gross columns are objects, while they could supposedly by numerical values. Let's have a deeper look at Gross. Since on the head there are no values for this column, we don't really know what is its formatting and if it needs any additional work.

In [10]:
movies_raw_dataset.loc[~movies_raw_dataset['Gross'].isna(), 'Gross'].head(20)

77      $75.47M
85     $402.45M
95      $89.22M
111    $315.54M
125     $57.01M
128    $260.00M
132    $132.38M
143    $167.77M
144    $404.52M
145     $15.07M
156     $70.10M
159    $210.61M
161    $327.48M
165    $390.53M
171    $303.00M
172     $56.63M
175     $58.06M
181    $353.01M
189     $46.89M
191      $7.00M
Name: Gross, dtype: object

So the Gross column will also need some formatting. 

We can now work on the column types and formatting

In [11]:
movies_raw_dataset.dtypes

Movies       object
Year         object
Genre        object
Rating      float64
One-line     object
Stars        object
Votes        object
Runtime     float64
Gross        object
dtype: object

In [12]:
movies_raw_dataset['Year']

0            (2021)
1          (2021– )
2       (2010–2022)
3          (2013– )
4            (2021)
           ...     
9994       (2021– )
9995       (2021– )
9996       (2022– )
9997       (2021– )
9998       (2021– )
Name: Year, Length: 9999, dtype: object

In [13]:
#Remove parenthesis from the Year Column
movies_raw_dataset['Year'] = movies_raw_dataset['Year'].str.replace(r'[()]','')

C:\Users\Hugo\AppData\Local\Temp\ipykernel_2604\546281310.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  movies_raw_dataset['Year'] = movies_raw_dataset['Year'].str.replace(r'[()]','')


In [14]:
#Remove new line char from Genre
movies_raw_dataset['Genre'] = movies_raw_dataset['Genre'].str.strip('\n')

In [15]:
#Remove new line char from One-Line
movies_raw_dataset['One-line'] = movies_raw_dataset['One-line'].str.strip('\n')

In [16]:
#Remove new line and special characters from Stars
movies_raw_dataset['Stars'] = movies_raw_dataset['Stars'].str.replace(r'[\n|]','')

C:\Users\Hugo\AppData\Local\Temp\ipykernel_2604\894442060.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  movies_raw_dataset['Stars'] = movies_raw_dataset['Stars'].str.replace(r'[\n|]','')


In [17]:
#Cast Votes as float to preserve the NaN. Converting to integer would make us have to fill those records and filling with 0 or a different number might skew future analysis
movies_raw_dataset['Votes'] = movies_raw_dataset['Votes'].str.replace(',','').astype(float)

In [18]:
#Get the Gross values correctly formatted
def number_formatter(x:str):
    '''Removes the letter identifier of the number magnitude and multiplies it by the value represented, returning a numerical variable'''
    if 'M' in x:
        return float(x.strip('M'))*1000000
    elif 'k' in x:
        return float(x.strip('k'))*1000
    else:
        return float(x)

In [19]:
movies_raw_dataset['Gross'] = movies_raw_dataset['Gross'].str.strip('$')
movies_raw_dataset['Gross'] = movies_raw_dataset['Gross'].apply(lambda x: number_formatter(str(x)))

In [20]:
movies_raw_dataset.head()

Movies       Year  \
0                        Blood Red Sky       2021   
1  Masters of the Universe: Revelation     2021–    
2                     The Walking Dead  2010–2022   
3                       Rick and Morty     2013–    
4                      Army of Thieves       2021   

                                      Genre  Rating  \
0      Action, Horror, Thriller                 6.1   
1  Animation, Action, Adventure                 5.0   
2       Drama, Horror, Thriller                 8.2   
3  Animation, Adventure, Comedy                 9.2   
4         Action, Crime, Horror                 NaN   

                                            One-line  \
0  A woman with a mysterious illness is forced in...   
1  The war for Eternia begins again in what may b...   
2  Sheriff Deputy Rick Grimes wakes up from a com...   
3  An animated series that follows the exploits o...   
4  A prequel, set before the events of Army of th...   

                                               Stars     Votes  Runtime  Gross  
0      Director:Peter Thorwarth     Stars:Peri Ba...   21062.0    121.0    NaN  
1                  Stars:Chris Wood, Sarah Michel...   17870.0     25.0    NaN  
2                  Stars:Andrew Lincoln, Norman R...  885805.0     44.0    NaN  
3                  Stars:Justin Roiland, Chris Pa...  414849.0     23.0    NaN  
4      Director:Matthias Schweighöfer     Stars:M...       NaN      NaN    NaN

Looking a lot better!

Let's now check for duplicates on the data.

In [21]:
#First we drop full duplicates
movies_raw_dataset.drop_duplicates(inplace=True, ignore_index=True)

In [22]:
#Checking for partial duplicates
movies_raw_dataset[movies_raw_dataset.duplicated(subset=['Movies','Year','Genre'], keep=False)].sort_values(by='Movies')

Movies       Year                                     Genre  \
8265    13 Reasons Why  2017–2020      Drama, Mystery, Thriller               
8109    13 Reasons Why  2017–2020      Drama, Mystery, Thriller               
8284    13 Reasons Why  2017–2020      Drama, Mystery, Thriller               
8108    13 Reasons Why  2017–2020      Drama, Mystery, Thriller               
8264    13 Reasons Why  2017–2020      Drama, Mystery, Thriller               
...                ...        ...                                       ...   
8340   ÜberWeihnachten       2020        Comedy, Drama, Romance               
8339   ÜberWeihnachten       2020        Comedy, Drama, Romance               
8338   ÜberWeihnachten       2020        Comedy, Drama, Romance               
6352           Far Cry        NaN  Animation, Action, Adventure               
5813           Far Cry        NaN  Animation, Action, Adventure               

      Rating                                           One-line  \
8265     5.9  Clay's mental health continues to decline as t...   
8109     6.6  The school goes into lockdown and Clay, Tony a...   
8284     6.8  When the dean begins a new investigation and t...   
8108     6.1  The Jensens make the boys take a drug test. Wh...   
8264     5.6  As the school gears up for the Love Is Love da...   
...      ...                                                ...   
8340     7.2  The Most Wonderful Christmas of the Year - the...   
8339     7.5  Sausages and Potato Salad - Basti hooks up wit...   
8338     7.2  Home Is Where the Tree Is - Bastian, an aspiri...   
6352     NaN  Plot under wraps. Adaptation of the Ubisoft game.   
5813     NaN  Feature film based on the video game 'Far Cry'...   

                                                  Stars   Votes  Runtime  \
8265      Director:Sunu Gonera     Stars:Dylan Minne...  1420.0     60.0   
8109      Director:Brenda Strong     Stars:Dylan Min...  1694.0     61.0   
8284      Director:Tommy Lohmann     Stars:Dylan Min...  1640.0     60.0   
8108      Director:Brenda Strong     Stars:Dylan Min...  1507.0     57.0   
8264      Director:Michael Sucsy     Stars:Dylan Min...  1635.0     59.0   
...                                                 ...     ...      ...   
8340                  Stars:Luke Mockridge, Seyneb S...    59.0     45.0   
8339                  Stars:Luke Mockridge, Seyneb S...    66.0     50.0   
8338                  Stars:Luke Mockridge, Seyneb S...    65.0     48.0   
6352                                                        NaN      NaN   
5813                                                        NaN      NaN   

      Gross  
8265    NaN  
8109    NaN  
8284    NaN  
8108    NaN  
8264    NaN  
...     ...  
8340    NaN  
8339    NaN  
8338    NaN  
6352    NaN  
5813    NaN  

[3025 rows x 9 columns]

It seems like there are series in which each line is an episode because they either have a different plot or a different director/actors. We can group this entries, using the rating and votes average, the sum of the runtime and concatenating the text fields to process later.

In [23]:
aggregations = {
    'Rating':'mean',
    'One-line':'sum',
    'Stars':'sum',
    'Votes':'mean',
    'Runtime':'mean',
    'Gross':'mean'
}

movies_raw_dataset = movies_raw_dataset.groupby(['Movies','Year','Genre']).agg(aggregations).reset_index()

Let's check again now

In [24]:
movies_raw_dataset[movies_raw_dataset.duplicated(subset=['Movies','Year','Genre'], keep=False)].sort_values(by='Movies')

Empty DataFrame
Columns: [Movies, Year, Genre, Rating, One-line, Stars, Votes, Runtime, Gross]
Index: []

Perfect! Our dataset looks much more clean!

Now it's time for testing if the data obeys to some defined rules. We have already checked that the rating is between the expected range, but now we need to check for the same thing in year, votes count and range.

To check if the year is in between boundries, let's create two new columns:
- **Starting year** - The year the movie was released or series started
- **Ending year** - The year series ended. In case of movies and series which were released during only one year, ending year shall be the same as starting year 

In [25]:
movies_raw_dataset['Startyear'] = movies_raw_dataset['Year'].apply(lambda x: x.split('–')[0])
movies_raw_dataset['Startyear'] = movies_raw_dataset['Startyear'].astype(int)

ValueError: invalid literal for int() with base 10: 'I'

Apprently some rows contain the value 'I' for the _Year_ column. Let's take a look at those rows.

In [26]:
movies_raw_dataset.loc[movies_raw_dataset['Year']=='I', :]

Movies Year  \
221                          Painkiller    I   
1179                            Bubbles    I   
1601                     Dead of Winter    I   
1903                            Empress    I   
2813                              Janet    I   
2938                         Juneteenth    I   
3362                                Lou    I   
4087                         Paranormal    I   
5010                   Sword Art Online    I   
5075                        Tell No One    I   
5402                         The Killer    I   
5526                        The Outlaws    I   
5631                       The Standoff    I   
5998  Untitled Thai Cave Rescue Project    I   
6109                          Watergate    I   

                                     Genre  Rating  \
221               Crime, Drama                 NaN   
1179         Animation, Comedy                 NaN   
1601                     Drama                 NaN   
1903  Action, Adventure, Drama                 NaN   
2813          Biography, Drama                 NaN   
2938                   Musical                 NaN   
3362     Crime, Drama, Mystery                 NaN   
4087    Drama, Horror, Mystery                 NaN   
5010  Action, Adventure, Drama                 NaN   
5075                  Thriller                 NaN   
5402      Action, Crime, Drama                 NaN   
5526                    Comedy                 NaN   
5631                     Drama                 NaN   
5998                     Drama                 NaN   
6109      Action, Crime, Drama                 NaN   

                                               One-line  \
221                                Add a PlotAdd a Plot   
1179  A chimpanzee tells his story of living with th...   
1601  An Asian American family's weekend in the moun...   
1903  The wife of a galactic dictator makes the deci...   
2813  A young black woman at the Washington Post and...   
2938                    A musical about Juneteenth, the   
3362  A tough elderly woman living on a remote islan...   
4087  Depicts the adventures of Dr. Refaat Ismail, a...   
5010  The series will follow a brilliant young beta ...   
5075  A widowed husband learns that his wife may sti...   
5402                                      Plot unknown.   
5526  Owen Browning is a bank manager about to marry...   
5631  Film takes place in December, 1969, when a new...   
5998  A boys soccer team in Thailand becomes trapped...   
6109  A look at the events and historical figures su...   

                                                  Stars  Votes  Runtime  \
221       Director:Peter Berg     Stars:Uzo Aduba, M...    NaN      NaN   
1179      Director:Mark Gustafson     Star:Piyali Mu...    NaN      NaN   
1601                                  Director:Mike Gan    NaN      NaN   
1903                                                       NaN      NaN   
2813                                Director:Janet Mock    NaN      NaN   
2938                                                       NaN      NaN   
3362      Director:Anna Foerster     Stars:Jurnee Sm...    NaN      NaN   
4087                                                       NaN      NaN   
5010                                                       NaN      NaN   
5075                           Director:Guillaume Canet    NaN      NaN   
5402      Director:David Fincher     Star:Michael Fa...    NaN      NaN   
5526      Director:Tyler Spindel     Stars:Pierce Br...    NaN      NaN   
5631                                Director:Justin Lin    NaN      NaN   
5998               Directors:Jon M. Chu, Baz Poonpiriya    NaN      NaN   
6109                                                       NaN      NaN   

      Gross Startyear  
221     NaN         I  
1179    NaN         I  
1601    NaN         I  
1903    NaN         I  
2813    NaN         I  
2938    NaN         I  
3362    NaN         I  
4087    NaN         I  
5010    NaN         I  
5075    NaN  

These movies do not have much info, so it seems like the I stands for Incoming. Let's see which rows contain anything but numbers and hyphen

In [27]:
movies_raw_dataset.loc[(movies_raw_dataset['Year'].str.contains('[a-zA-Z]')), :]

Movies             Year  \
221                                          Painkiller                I   
314                                     The Pentaverate               II   
354                                          True Story              III   
413   13th: A Conversation with Oprah Winfrey & Ava ...       2017 Video   
423                                                1994         I 2019–    
...                                                 ...              ...   
6319                                Zeitgeist: Addendum       2008 Video   
6334                                               Zion           I 2018   
6336                                                Zoe           I 2018   
6342                             Zoolander: Super Model    2016 TV Movie   
6357                                       Året der gik  2020 TV Special   

                                          Genre  Rating  \
221                    Crime, Drama                 NaN   
314                          Comedy                 NaN   
354                   Comedy, Drama                 NaN   
413              Documentary, Short                 7.1   
423            Documentary, History                 7.6   
...                                         ...     ...   
6319      Documentary, History, War                 8.2   
6334  Documentary, Short, Biography                 7.2   
6336                Romance, Sci-Fi                 6.1   
6342      Animation, Comedy, Sci-Fi                 4.9   
6357      History, Music, Talk-Show                 NaN   

                                               One-line  \
221                                Add a PlotAdd a Plot   
314   Add a PlotAdd a PlotAdd a PlotAdd a PlotAdd a ...   
354            Add a PlotAdd a PlotAdd a PlotAdd a Plot   
413   Oprah Winfrey sits down with other to discuss ...   
423   Archival video and new interviews provide an i...   
...                                                 ...   
6319  Zeitgeist: Addendum attempts to locate the roo...   
6334  A portrait of Zion Clark, a young wrestler who...   
6336  A story about how synthetic humans can feel an...   
6342  In this animated serialized movie, ridiculousl...   
6357                                         Add a Plot   

                                                  Stars    Votes  Runtime  \
221       Director:Peter Berg     Stars:Uzo Aduba, M...      NaN      NaN   
314       Director:Tim Kirkby     Star:Phill Webster...      NaN      NaN   
354       Director:Stephen Williams     Stars:Kevin ...      NaN      NaN   
413                   Stars:Oprah Winfrey, Ava DuVernay    193.0     37.0   
423                   Stars:Rafael Sebastián Guillén...    442.0      NaN   
...                                                 ...      ...      ...   
6319      Director:Peter Joseph     Stars:Jacque Fre...  18071.0    123.0   
6334      Director:Floyd Russ     Stars:Zion Clark, ...   1096.0     12.0   
6336      Director:Drake Doremus     Stars:Ewan McGr...   8146.0    104.0   
6342      Director:Aaron Augenblick     Stars:Ben St...    139.0     84.0   
6357                  Stars:Stéphanie Surrugue, Kåre...      NaN    134.0   

      Gross        Startyear  
221     NaN                I  
314     NaN               II  
354     NaN              III  
413     NaN       2017 Video  
423     NaN           I 2019  
...     ...              ...  
6319    NaN       2008 Video  
6334    NaN           I 2018  
6336    NaN           I 2018  
6342    NaN    2016 TV Movie  
6357    NaN  2020 TV Special  

[1003 rows x 10 columns]

A lot of entries with more information than the year. Let's look at the unique values on these rows.

In [28]:
movies_raw_dataset.loc[(movies_raw_dataset['Year'].str.contains('[a-zA-Z]')), 'Year'].unique()

array(['I', 'II', 'III', '2017 Video', 'I 2019– ', '2020 TV Special',
       'I 2017', '2014 TV Special', '2019 Video', '2013 Video',
       '2008 Video', '2019 TV Movie', '2017 TV Movie', '2014 Video',
       '2015 TV Special', '2016 TV Movie', '2019 TV Special',
       '2018 TV Special', '2017 TV Special', 'III 2021', 'IV 2021',
       '2020 TV Movie', '2021 TV Special', '2016 TV Special',
       '2006 Video', '2014 TV Movie', '2006 TV Movie', 'I 2021',
       'II 2021', 'I 2014', 'TV Special', '2012 Video', 'I 2019',
       '2021 TV Movie', 'I 2018', '2015 Video', 'I 2012', '2011 Video',
       'II 2016', 'IV 2018', 'I 2009', 'I 2016', '2013 TV Special',
       '2011 TV Movie', 'II 2017', 'II 2015', 'I 2020', 'I 2008',
       '2021 Video', 'II 2020', 'II 2013', '2012 TV Special',
       '1992 Video', '1993 TV Special', '1989 Video', '2013 TV Movie',
       'XLI', 'II 2022', 'II 2009', 'I 2020– ', 'I 2011', '2016 Video',
       'IV 2020', 'IV', 'III 2015', '2011 TV Special', 'III 202

Well, at least most of them have the year info, so we will be able to extract it somehow. For now, let's take a look on the ones who do not and see if they are interesting to keep.

In [29]:
no_year_info = movies_raw_dataset.loc[~(movies_raw_dataset['Year'].str.contains('[0-9]')), :]

In [30]:
no_year_info.describe()

Rating  Votes     Runtime  Gross
count     0.0    0.0    2.000000    0.0
mean      NaN    NaN   94.000000    NaN
std       NaN    NaN   48.083261    NaN
min       NaN    NaN   60.000000    NaN
25%       NaN    NaN   77.000000    NaN
50%       NaN    NaN   94.000000    NaN
75%       NaN    NaN  111.000000    NaN
max       NaN    NaN  128.000000    NaN

Do not seem to be very relevant, as there's barely any information on _Rating_ and _Votes_. Let's drop this from the dataset.

In [31]:
movies_raw_dataset = movies_raw_dataset.loc[(movies_raw_dataset['Year'].str.contains('[0-9]')), :]

From here we can filter only the digits from the _year_ column

In [32]:
movies_raw_dataset['Year'] = movies_raw_dataset['Year'].str.replace('[^0-9\–]','')

C:\Users\Hugo\AppData\Local\Temp\ipykernel_2604\1470456750.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  movies_raw_dataset['Year'] = movies_raw_dataset['Year'].str.replace('[^0-9\–]','')


And now we can proceed as intended

In [33]:
movies_raw_dataset['Year'] = movies_raw_dataset['Year'].str.strip('–')

In [34]:
movies_raw_dataset['Startyear'] = movies_raw_dataset['Year'].apply(lambda x: x.split('–')[0])
movies_raw_dataset['Startyear'] = movies_raw_dataset['Startyear'].astype(int)

In [35]:
movies_raw_dataset['Endyear'] = movies_raw_dataset['Year'].apply(lambda x: x.split('–')[-1])
movies_raw_dataset['Endyear'] = movies_raw_dataset['Endyear'].astype(int)

Check again on our numeric variables

In [36]:
movies_raw_dataset.describe()

Rating         Votes      Runtime         Gross    Startyear  \
count  5845.000000  5.845000e+03  5320.000000  4.600000e+02  6297.000000   
mean      6.637888  2.086695e+04    78.719302  4.370187e+07  2015.560108   
std       1.215359  8.216492e+04    49.783835  8.242330e+07     8.145803   
min       1.100000  5.000000e+00     1.000000  0.000000e+00  1932.000000   
25%       5.900000  3.480000e+02    45.000000  1.500000e+05  2015.000000   
50%       6.800000  1.478000e+03    82.000000  6.145000e+06  2018.000000   
75%       7.500000  6.943000e+03   100.000000  4.694750e+07  2019.000000   
max       9.400000  1.713028e+06   853.000000  5.040100e+08  2023.000000   

           Endyear  
count  6297.000000  
mean   2015.899158  
std       8.016333  
min    1932.000000  
25%    2015.000000  
50%    2018.000000  
75%    2020.000000  
max    2023.000000

Seems a bit odd that we have a runtime of 1 minutes. Let's take a look at the rows with the smaller runtimes.

In [37]:
movies_raw_dataset.nsmallest(10, 'Runtime')

Movies       Year  \
353                                        Trollhunters  2016–2018   
411   13 Reasons Why: Season 2 Date Announcement Com...       2018   
658        All Hail King Julien: Happy Birthday to You!       2017   
906                      Barbie: Happy Birthday to You!       2017   
930                   Beat Bugs: Happy Birthday to You!       2017   
1695                   Dinotrux: Happy Birthday to You!       2017   
2588                                     Horsin' Around       2014   
3118               LEGO Friends: Happy Birthday to You!       2017   
3119  LEGO Ninjago: Masters of Spinjitzu: Happy Birt...       2017   
3627  Miraculous: Ladybug & Cat Noir: Happy Birthday...       2017   

                                         Genre  Rating  \
353   Animation, Action, Adventure                 8.2   
411                   Short, Drama                 7.2   
658               Animation, Short                 4.4   
906               Animation, Short                 4.1   
930               Animation, Short                 NaN   
1695              Animation, Short                 NaN   
2588          Short, Comedy, Music                 7.4   
3118              Animation, Short                 3.6   
3119      Animation, Short, Action                 5.9   
3627              Animation, Short                 7.3   

                                               One-line  \
353   Trolls, goblins, humans, Gumm-Gumms -- all man...   
411                                          Add a Plot   
658   Nobody parties like King Julien. And today he'...   
906   Get ready to make a wish and have the most ama...   
930   They say it's your birthday - and there's no o...   
1695  Ty and the rest of the Dinotrux just found out...   
2588  The opening title sequence to the spoof 90s si...   
3118  Surprise. The gang's all here to celebrate you...   
3119  LEGO Ninjago heroes Kai, Zane, Jay and Cole wa...   
3627  Break dancing, dragons, balloons and a DJ: Lad...   

                                                  Stars  Votes  Runtime  \
353                                                       16.5      1.0   
411       Director:Greg Jardin     Stars:Alisha Boe,...  197.0      1.0   
658                                                        7.0      1.0   
906                   Stars:Ritesh Rajan, America Young    8.0      1.0   
930                                                        NaN      1.0   
1695                                                       NaN      1.0   
2588                                                     627.0      1.0   
3118                                                       5.0      1.0   
3119                                                      10.0      1.0   
3627                                                      34.0      1.0   

      Gross  Startyear  Endyear  
353     NaN       2016     2018  
411     NaN       2018     2018  
658     NaN       2017     2017  
906     NaN       2017     2017  
930     NaN       2017     2017  
1695    NaN       2017     2017  
2588    NaN       2014     2014  
3118    NaN       2017     2017  
3119    NaN       2017     2017  
3627    NaN       2017     2017

Most of these are classified as shorts, so it is actually possible that the value is correct.

In [38]:
movies_raw_dataset.head()

Movies       Year  \
0             13 Reasons Why  2017–2020   
1                       1899       2022   
2   3Below: Tales of Arcadia  2018–2019   
3                       50M2       2021   
4                     7Seeds  2019–2020   

                                      Genre    Rating  \
0      Drama, Mystery, Thriller              6.140000   
1        Drama, History, Horror                   NaN   
2  Animation, Action, Adventure              7.950000   
3       Comedy, Drama, Thriller              7.300000   
4  Animation, Action, Adventure              7.108696   

                                            One-line  \
0  The police question Tyler about the guns, leav...   
1                               Add a PlotAdd a Plot   
2  Left vulnerable after Omen's attack, the royal...   
3  While seeking answers about his parents, Shado...   
4  Convinced that Botan has kidnapped them, Natsu...   

                                               Stars        Votes    Runtime  \
0      Director:Russell Mulcahy     Stars:Dylan M...  1693.600000  62.800000   
1      Director:Baran bo Odar     Stars:Aneurin B...          NaN        NaN   
2      Director:Andrew L. Schmidt     Stars:Tatia...   129.500000  20.461538   
3      Director:Selçuk Aydemir     Stars:Engin Öz...   119.500000  49.000000   
4                  Stars:Morgan Berry, Amber Lee ...    30.521739  24.695652   

   Gross  Startyear  Endyear  
0    NaN       2017     2020  
1    NaN       2022     2022  
2    NaN       2018     2019  
3    NaN       2021     2021  
4    NaN       2019     2020

Isolating the Director might be worthy to use it in an analysis later. However, we still have some issues here that we need to be careful with:
- Director's names are likely standardized due to the fact that IMDB has a profile for each entity. Still, we cannot take this as guaranteed without analyzing further.
- Some series have multiple directors, as they might have different directors for different episodes
- A large number of directors might compromise the resources necessary to develop a model, as it would create a large number of new columns for the data frame.

In [39]:
movies_raw_dataset['Director'] = movies_raw_dataset['Stars'].apply(lambda x: x[x.find('Director:')+9:x.find('Stars:')].strip())

In [133]:
#Regex with (Director:|Directors:) was not working, so had to work another way

def extract_director(stars_string):

    stars_string = stars_string.strip()

    while 'Stars:' in stars_string:
        index_stars = stars_string.find('Stars:')
        index_director = stars_string[index_stars:].find('Director')
        if index_director != -1:
            global_index = index_director + index_stars
            stars_string = stars_string[0:index_stars]+stars_string[global_index:]
        else:
            stars_string = stars_string[0:index_stars]

    stars_string = stars_string.strip()
    stars_string = stars_string.replace('Directors:', '|')
    stars_string = stars_string.replace('Director:', '|')

    directors_list = []
    
    for s in stars_string.split('|'):
        for t in s.split(','):
            t = t.strip('\n\t ')
            if t in directors_list or t == '':
                continue
            else: 
                directors_list.append(t)

    directors_string = ','.join(directors_list)
    return directors_string


In [136]:
movies_raw_dataset['Director'] = movies_raw_dataset['Stars'].apply(lambda x: extract_director(x))

Director:Russell Mulcahy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Russell Mulcahy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Michael Sucsy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Sunu Gonera     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brenda Strong     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brenda Strong     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Michael Sucsy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Sunu Gonera     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Tommy Lohmann     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brian Yorkey     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe
Director:Russell Mulcahy     Director:Russell Mulcahy     Stars:

In [134]:
extract_director(movies_raw_dataset['Stars'][0])

Director:Russell Mulcahy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Russell Mulcahy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Michael Sucsy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Sunu Gonera     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brenda Strong     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brenda Strong     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Michael Sucsy     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Sunu Gonera     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Tommy Lohmann     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe    Director:Brian Yorkey     Stars:Dylan Minnette, Grace Saif, Christian Navarro, Alisha Boe
Director:Russell Mulcahy     Director:Russell Mulcahy     Stars:

'Russell Mulcahy,Michael Sucsy,Sunu Gonera,Brenda Strong,Tommy Lohmann,Brian Yorkey'

In [122]:
movies_raw_dataset.loc[movies_raw_dataset['Director'].str.contains(':'),:]

Movies       Year  \
0                         13 Reasons Why  2017–2020   
3                                   50M2       2021   
6                         A Suitable Boy       2020   
8                             After Life       2019   
9       Age of Samurai: Battle for Japan       2021   
...                                  ...        ...   
6281                           Yoga Time  2004–2009   
6284       Yoo Byungjae: Discomfort Zone       2018   
6285  Yoo Byungjae: Too Much Information       2018   
6293                        You vs. Wild       2019   
6298         Yours Sincerely, Kanan Gill       2020   

                             Genre    Rating  \
0           Drama,Mystery,Thriller  6.140000   
3            Comedy,Drama,Thriller  7.300000   
6                    Drama,Romance  6.200000   
8                     Comedy,Drama  8.283333   
9     Documentary,Action,Adventure  7.450000   
...                            ...       ...   
6281                        Family       NaN   
6284                        Comedy  4.800000   
6285                        Comedy  5.300000   
6293          Adventure,Reality-TV  6.700000   
6298                        Comedy  7.300000   

                                               One-line  \
0     The police question Tyler about the guns, leav...   
3     While seeking answers about his parents, Shado...   
6     Lata continues to pursue a relationship with K...   
8     Determined to be kind but still struggling to ...   
9     Considered a fool and unfit to lead, Nobunaga ...   
...                                                 ...   
6281  Instructional Hatha yoga show teaching posture...   
6284                                                      
6285  Korean stand-up comic Yoo Byungjae offers his ...   
6293  In this interactive series, you'll make key de...   
6298  Revisiting life goals set in a letter written ...   

                                                  Stars        Votes  \
0         Director:Russell Mulcahy     Stars:Dylan M...  1693.600000   
3         Director:Selçuk Aydemir     Stars:Engin Öz...   119.500000   
6         Director:Mira Nair     Stars:Tanya Manikta...   148.833333   
8         Director:Ricky Gervais     Stars:Ricky Ger...  1836.833333   
9         Director:Stephen Scott     Stars:Hiro Kana...   166.000000   
...                                                 ...          ...   
6281                                Star:Marlon Braccia          NaN   
6284       Director:Y. Joon Chung     Star:Byungjae Yoo    25.000000   
6285        Director:Byungjae Yoo     Star:Byungjae Yoo    49.000000   
6293                                   Star:Bear Grylls  2333.000000   
6298          Director:Neville Shah     Star:Kanan Gill   828.000000   

        Runtime  Gross  Startyear  Endyear  \
0     62.800000    NaN       2017     2020   
3     49.000000    NaN       2021     2021   
6     58.000000    NaN       2020     2020   
8     28.166667    NaN       2019     2019   
9     42.833333    NaN       2021     2021   
...         ...    ...        ...      ...   
6281  28.000000    NaN       2004     2009   
6284  54.000000    NaN       2018     2018   
6285  63.000000    NaN       2018     2018   
6293  20.000000    NaN       2019     2019   
6298  72.000000    NaN       2020     2020   

                                               Director  
0     or:Russell Mulcahy,Russell Mulcahy,Michael Suc...  
3        ctor:Selçuk Aydemir,Selçuk Aydemir,Burak Aksak  
6                rector:Mira Nair,Mira Nair,Shimit Amin  
8                    rector:Ricky Gervais,Ricky Gervais  
9                    rector:Stephen Scott,Stephen Scott  
...                                                 ...  
6281                                Star:Marlon Braccia  
6284                Y. Joon Chung     Star:Byungjae Yoo  
6285                 Byungjae Yoo     Star:Byungjae Yoo  
6293                                   Star:Bear Grylls  
6298                   Neville 

Everything seems fine with the names extracted. 

In [44]:
movies_raw_dataset.loc[movies_raw_dataset['Director']!='','Director'].nunique()

3277

So many directors seem to be too much for creating dummy variables for them, so let's leave it for now. We keep the stars as a whole in case it is necessary to figure out if a certain actor participated on the movie or series.

Let's remove the One-liners with no plot

In [45]:
movies_raw_dataset['One-line'] = movies_raw_dataset['One-line'].str.replace('Add a Plot','')

Let's also create dummies for the movie genre. 

In [46]:
movies_raw_dataset['Genre'] = movies_raw_dataset['Genre'].str.replace(' ','')
genre_dummies = movies_raw_dataset['Genre'].str.get_dummies(',')

movies_dataset = pd.concat([movies_raw_dataset, genre_dummies], axis=1)
movies_dataset.drop('Genre', axis=1, inplace=True)

In [ ]:
movies_dataset.head()

Movies       Year    Rating  \
0             13 Reasons Why  2017–2020  6.140000   
1                       1899       2022       NaN   
2   3Below: Tales of Arcadia  2018–2019  7.950000   
3                       50M2       2021  7.300000   
4                     7Seeds  2019–2020  7.108696   

                                            One-line  \
0  The police question Tyler about the guns, leav...   
1                                                      
2  Left vulnerable after Omen's attack, the royal...   
3  While seeking answers about his parents, Shado...   
4  Convinced that Botan has kidnapped them, Natsu...   

                                               Stars        Votes    Runtime  \
0      Director:Russell Mulcahy     Stars:Dylan M...  1693.600000  62.800000   
1      Director:Baran bo Odar     Stars:Aneurin B...          NaN        NaN   
2      Director:Andrew L. Schmidt     Stars:Tatia...   129.500000  20.461538   
3      Director:Selçuk Aydemir     Stars:Engin Öz...   119.500000  49.000000   
4                  Stars:Morgan Berry, Amber Lee ...    30.521739  24.695652   

   Gross  Startyear  Endyear  ... News  Reality-TV  Romance  Sci-Fi  Short  \
0    NaN       2017     2020  ...    0           0        0       0      0   
1    NaN       2022     2022  ...    0           0        0       0      0   
2    NaN       2018     2019  ...    0           0        0       0      0   
3    NaN       2021     2021  ...    0           0        0       0      0   
4    NaN       2019     2020  ...    0           0        0       0      0   

   Sport  Talk-Show  Thriller  War  Western  
0      0          0         1    0        0  
1      0          0         0    0        0  
2      0          0         0    0        0  
3      0          0         1    0        0  
4      0          0         0    0        0  

[5 rows x 38 columns]

Ideas not implemented:
- We can try to divided the dataset into a movies dataset and a series dataset. However, our only criteria would have to be the presence of multiple values in the _Year_ column, which would mean that the series went on for several years. This is true for most series, but for series broadcasted within the span of a single year, that classification would be erroneous.


Let's export the data to a .parquet file to access it later.

In [ ]:
movies_dataset.to_parquet('movies_dataset_clean.parquet')